src/ratelimiter.rs

```diff
@@ -0,0 +1,45 @@
+use std::thread::sleep;
+use std::time::{Duration, Instant};
+
+#[derive(Debug)]
+pub struct RateLimiter {
+    available_tokens: f32,
+    tokens_per_second: f32,
+    last_update: Instant,
+}
+
+impl RateLimiter {
+    pub(crate) fn new(chunk_size_bytes: u64, mbps_target: Option<u64>) -> Self {
+        let bps_target = mbps_target.map(|x| x * 1_000_000);
+        let tokens_per_second = match bps_target {
+            None => 0.0,
+            Some(bps) => bps as f32 / chunk_size_bytes as f32,
+        };
+        RateLimiter {
+            available_tokens: 0.0,
+            tokens_per_second,
+            last_update: Instant::now(),
+        }
+    }
+
+    pub(crate) fn block_until_available(&mut self) {
```

Comment:
I would make the interface something like this:

```rust
pub fn bytes_available(&self, now: Instant) -> f32;
pub fn consume_bytes(&mut self, now: Instant, num_bytes: f32);
pub fn time_until_bytes_available(&self, now: Instant, num_bytes: f32) -> Duration;
```

Passing in `now` externally is not strictly needed but it makes the entire thing pure which makes it easier to test because you can say do something like this:

```rust
let mut rl = RateLimiter::new(...);
let t0 = Instant::now();
t1 = t0 + rl.time_until_available_bytes(t0, 100.0);
assert!(rl.bytes_available(t1) >= 100.0);
```
without the test having to block for a long time.

Reply:
This is a lot nicer, though I did not end up using `bytes_available` within `fastsync`, only to test internal behavior